From 3762e7be0368484d68c4a90923672fdb0f51a5fc Mon Sep 17 00:00:00 2001
From: Patrick Zwerschke <patrick@p5k.org>
Date: Tue, 6 Dec 2022 03:41:27 +0100
Subject: [PATCH] Update test workflow for python 3.11 (#215)

- added python 3.11 to test workflow
- catch FileNotFoundError in module defaultpaths
---
 .github/workflows/pytest.yml           | 10 +++++++---
 DisplayCAL/RealDisplaySizeMM.py        |  2 ++
 DisplayCAL/defaultpaths.py             | 17 +++++++++--------
 DisplayCAL/lib64/python311/__init__.py |  0
 README.md                              |  2 +-
 5 files changed, 19 insertions(+), 12 deletions(-)
 create mode 100644 DisplayCAL/lib64/python311/__init__.py

diff --git a/DisplayCAL/RealDisplaySizeMM.py b/DisplayCAL/RealDisplaySizeMM.py
index 00957692c..25b730d05 100644
--- a/DisplayCAL/RealDisplaySizeMM.py
+++ b/DisplayCAL/RealDisplaySizeMM.py
@@ -25,6 +25,8 @@
         from DisplayCAL.lib64.python39.RealDisplaySizeMM import *
     elif sys.version_info[:2] == (3, 10):
         from DisplayCAL.lib64.python310.RealDisplaySizeMM import *
+    elif sys.version_info[:2] == (3, 11):
+        from DisplayCAL.lib64.python311.RealDisplaySizeMM import *
 # else:
 #     pass
 
diff --git a/DisplayCAL/defaultpaths.py b/DisplayCAL/defaultpaths.py
index cd11d6d75..59258e79d 100644
--- a/DisplayCAL/defaultpaths.py
+++ b/DisplayCAL/defaultpaths.py
@@ -235,9 +235,14 @@ def set_translation(obj):
                     obj.GETTEXT_PACKAGE, locale_dir, codeset="UTF-8"
                 )
             except TypeError:
-                obj.translation = gettext.translation(
-                    obj.GETTEXT_PACKAGE, locale_dir
-                )
+                try:
+                    obj.translation = gettext.translation(
+                        obj.GETTEXT_PACKAGE, locale_dir
+                    )
+                except FileNotFoundError as exc:
+                    print("XDG:", exc)
+                    obj.translation = gettext.NullTranslations()
+                    return False
             except IOError as exception:
                 print("XDG:", exception)
                 obj.translation = gettext.NullTranslations()
@@ -261,11 +266,7 @@ def get_config_files(filename):
 
         @staticmethod
         def shell_unescape(s):
-            a = []
-            for i, c in enumerate(s):
-                if c == "\\" and len(s) > i + 1:
-                    continue
-                a.append(c)
+            a = [c for i, c in enumerate(s) if c != "\\" or len(s) <= i + 1]
             return "".join(a)
 
         @staticmethod
diff --git a/DisplayCAL/lib64/python311/__init__.py b/DisplayCAL/lib64/python311/__init__.py
new file mode 100644
index 000000000..e69de29bb
diff --git a/README.md b/README.md
index 3ff3c4ec4..4119ace70 100644
--- a/README.md
+++ b/README.md
@@ -25,7 +25,7 @@ Here is a screenshots showing the tool working with Python 3.10:
 
 ![image](https://user-images.githubusercontent.com/1786804/169152229-e06ff549-55fe-4149-8742-405446e6b01f.png)
 
-Currently, DisplayCAL is working with Python 3.8, 3.9 and 3.10 and wxPython 4.1.1 or 4.2.0.
+Currently, DisplayCAL is working with Python 3.8, 3.9, 3.10 and 3.11 and wxPython 4.1.1 or 4.2.0.
 
 Here is a list of things that is working:
 
